collaborative-plonk/src/round2.rs

```diff
@@ -0,0 +1,341 @@
+            z,
+        }
+    }
+}
+
+impl<T, P: Pairing> Round2<T, P>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>
+        + PairingEcMpcProtocol<P>
+        + FFTProvider<P::ScalarField>
+        + MSMProvider<P::G1>
+        + MSMProvider<P::G2>,
+    P::ScalarField: FFTPostProcessing,
+{
+    fn compute_z(
+        driver: &mut T,
+        zkey: &ZKey<P>,
+        domains: &Domains<P>,
+        challenges: &Round2Challenges<T, P>,
+        polys: &Round1Polys<T, P>,
+    ) -> PlonkProofResult<PolyEval<T, P>> {
+        // TODO Check if this root_of_unity is the one we need
+        // FIXME  Do we want the dependency to collaborative Groth16??
+        let pow_root_of_unity = domains.roots_of_unity[zkey.power];
+        let mut w = P::ScalarField::one();
+        let mut n1 = Vec::with_capacity(zkey.domain_size);
+        let mut n2 = Vec::with_capacity(zkey.domain_size);
+        let mut n3 = Vec::with_capacity(zkey.domain_size);
+        let mut d1 = Vec::with_capacity(zkey.domain_size);
+        let mut d2 = Vec::with_capacity(zkey.domain_size);
+        let mut d3 = Vec::with_capacity(zkey.domain_size);
+        for i in 0..zkey.domain_size {
+            let a = T::index_sharevec(&polys.buffer_a, i);
+            let b = T::index_sharevec(&polys.buffer_b, i);
+            let c = T::index_sharevec(&polys.buffer_c, i);
+
+            // Z(X) := numArr / denArr
+            // numArr := (a + beta·ω + gamma)(b + beta·ω·k1 + gamma)(c + beta·ω·k2 + gamma)
+            let betaw = challenges.beta * w;
+
+            let n1_ = driver.add_with_public(&betaw, &a);
+            let n1_ = driver.add_with_public(&challenges.gamma, &n1_);
+
+            let tmp = zkey.verifying_key.k1 * betaw;
+            let n2_ = driver.add_with_public(&tmp, &b);
+            let n2_ = driver.add_with_public(&challenges.gamma, &n2_);
+
+            let tmp = zkey.verifying_key.k2 * betaw;
+            let n3_ = driver.add_with_public(&tmp, &c);
+            let n3_ = driver.add_with_public(&challenges.gamma, &n3_);
+
+            n1.push(n1_);
+            n2.push(n2_);
+            n3.push(n3_);
+
+            // denArr := (a + beta·sigma1 + gamma)(b + beta·sigma2 + gamma)(c + beta·sigma3 + gamma)
+            let d1_ =
+                driver.add_with_public(&(challenges.beta * zkey.s1_poly.evaluations[i * 4]), &a);
+            let d1_ = driver.add_with_public(&challenges.gamma, &d1_);
+
+            let d2_ =
+                driver.add_with_public(&(challenges.beta * zkey.s2_poly.evaluations[i * 4]), &b);
+            let d2_ = driver.add_with_public(&challenges.gamma, &d2_);
+
+            let d3_ =
+                driver.add_with_public(&(challenges.beta * zkey.s3_poly.evaluations[i * 4]), &c);
+            let d3_ = driver.add_with_public(&challenges.gamma, &d3_);
+
+            d1.push(d1_);
+            d2.push(d2_);
+            d3.push(d3_);
+
+            w *= &pow_root_of_unity;
+        }
+
+        // TODO parallelize these?
```

Comment:
This is something a generically better computing/networking driver would then do for us later

Reply:
so let's keep the comment in here?